### ISSUE TITLE
Improvements to current pub IP logic

### ISSUE THREAD

**Issue by DSdatsme** [HarshKapadia2/dynamic-dns#1]
Following are the improvements you could consider for updating the [DDNS script](https://github.com/HarshKapadia2/dynamic-dns/blob/main/ddns).

1. Handling no current file present
https://github.com/HarshKapadia2/dynamic-dns/blob/main/ddns#L19-L21 
**Resolution**: There is no point in proceeding further if the file public-ip-in-use if absent. The if condition should have an `exit 1` statement.

2. Logic in updating public-ip-in-use
As soon as the program sees the IP as changed it tries to update the local state first without considering whether the remote state has been updated.
https://github.com/HarshKapadia2/dynamic-dns/blob/main/ddns#L25
**Resolution**: Move the update IP to text file line inside the if condition `"${cloudflare_http_res_code}" == "200"`
This handles the scenario when Cloudflare is down, API key errors, the internet not working, etc .. and your IP on local text file is already updated on next run. 

3. Doc/automation improvements for configurations
- How about storing all env variables like token,zone id, etc in a `.local_ddns` file in `/home/<user>` dir. when the script starts it'll load up using commands like `source` .
- Create a script (install.sh) that can directly add required cron, populate env file (run-ddns, list-all-dns-records) , everything at one go, making setup process hassle free.

**Comment by HarshKapadia2** [comment]
Thank you so much for this feedback, Darshit! I really appreciate it!

> 1. Handling no current file present
>    [`main`/ddns#L19-L21](https://github.com/HarshKapadia2/dynamic-dns/blob/main/ddns?rgh-link-date=2023-09-10T07%3A48%3A53Z#L19-L21)
>    **Resolution**: There is no point in proceeding further if the file public-ip-in-use if absent. The if condition should have an `exit 1` statement.

If the file doesn't exist, then [it will be created later on](https://github.com/HarshKapadia2/dynamic-dns/blob/617f54c783654f6b7b6c7a45891c25b0aa3150b7/ddns#L25), so there is no need to stop the script here.

(Also, when the `ddns` script is run for the first time, the `public-ip-in-use.txt` file will not exist.)

> 2. Logic in updating public-ip-in-use
>    As soon as the program sees the IP as changed it tries to update the local state first without considering whether the remote state has been updated.
>    [`main`/ddns#L25](https://github.com/HarshKapadia2/dynamic-dns/blob/main/ddns?rgh-link-date=2023-09-10T07%3A48%3A53Z#L25)
>    **Resolution**: Move the update IP to text file line inside the if condition `"${cloudflare_http_res_code}" == "200"`
>    This handles the scenario when Cloudflare is down, API key errors, the internet not working, etc .. and your IP on local text file is already updated on next run.

Ah yes, good catch. This makes a lot of sense. I'll make this change. Thank you!

With the current implementation, if the public IP changes and updates the stored file but fails to update Cloudflare, the new public IP will never make it to Cloudflare until the public IP changes again, which might happen after a lot of days.

> 3. Doc/automation improvements for configurations
> 
> * How about storing all env variables like token,zone id, etc in a `.local_ddns` file in `/home/<user>` dir. when the script starts it'll load up using commands like `source` .

I don't understand how I would get the variables in the `.local_ddns` file using `source`. Do you mean reading the file and parsing variables? That's too complicated.

> * Create a script (install.sh) that can directly add required cron, populate env file (run-ddns, list-all-dns-records) , everything at one go, making setup process hassle free.

Ah yes, this will be nice to have, but I don't think I have the motivation to put in the time to do this, as I've already set everything up. Good suggestion though. Thank you!

Again, thank you so much for the great feedback @DSdatsme! This is good learning for me.

PS: @KartikSoneji also had [some feedback for me](https://github.com/HarshKapadia2/dynamic-dns/commit/67811943bdce1f5f0eb393781b04a7f61dd385f6) and you might want to have a look at it!

**Comment by HarshKapadia2** [comment]
Oh by the way @DSdatsme, it would be better to use permalinks to link to specific lines, as those lines might change over time and permalinks fix the commit that you were referring to, thus combating the change in content of those lines over time.

**Comment by KartikSoneji** [comment]
It might be better to read the current ip from cloudflare itself instead of storing it locally.
That way if they are ever out of sync for any reason, running the script will ensure everything is up to date.

**Comment by HarshKapadia2** [comment]
> It might be better to read the current ip from cloudflare itself instead of storing it locally.
> That way if they are ever out of sync for any reason, running the script will ensure everything is up to date.

Yeah that's a good idea, but now that this is working, I don't think I have the time to change it. Also, I tried parsing Curl's HTTP response JSON output using `jq`, but it will take me some time to figure out exactly how it works, so I'm going to leave it for now. Good idea though. Thank you for the suggestion, Kartik!

@KartikSoneji
